pkg/controller/smart-clone-controller.go

```diff
@@ -1,377 +1,234 @@
-	csiClientSet csiclientset.Interface,
-	pvcInformer coreinformers.PersistentVolumeClaimInformer,
-	snapshotInformer snapshotsinformers.VolumeSnapshotInformer,
-	dataVolumeInformer informers.DataVolumeInformer) *SmartCloneController {
-
-	// Create event broadcaster
-	// Add smart-clone-controller types to the default Kubernetes Scheme so Events can be
-	// logged for smart-clone-controller types.
-	cdischeme.AddToScheme(scheme.Scheme)
-	klog.V(3).Info("Creating event broadcaster")
-	eventBroadcaster := record.NewBroadcaster()
-	eventBroadcaster.StartLogging(klog.V(2).Infof)
-	eventBroadcaster.StartRecordingToSink(&typedcorev1.EventSinkImpl{Interface: client.CoreV1().Events("")})
-	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: controllerAgentName})
-
-	c := &SmartCloneController{
-		clientset:         client,
-		cdiClientSet:      cdiClientSet,
-		csiClientSet:      csiClientSet,
-		pvcLister:         pvcInformer.Lister(),
-		snapshotsLister:   snapshotInformer.Lister(),
-		snapshotInformer:  snapshotInformer.Informer(),
-		dataVolumeLister:  dataVolumeInformer.Lister(),
-		dataVolumesSynced: dataVolumeInformer.Informer().HasSynced,
-		snapshotsSynced:   snapshotInformer.Informer().HasSynced,
-		recorder:          recorder,
-		queue:             workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+// NewSmartCloneController creates a new instance of the import controller.
+func NewSmartCloneController(mgr manager.Manager, log logr.Logger) (controller.Controller, error) {
+	reconciler := &SmartCloneReconciler{
+		Client:   mgr.GetClient(),
+		Scheme:   mgr.GetScheme(),
+		Log:      log.WithName("smartclone-controller"),
+		recorder: mgr.GetEventRecorderFor("smartclone-controller"),
 	}
-
-	// Set up an event handler for when VolumeSnapshot resources change
-	snapshotInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
-		AddFunc: c.enqueueItem,
-		UpdateFunc: func(old, new interface{}) {
-			c.enqueueItem(new)
-		},
-		DeleteFunc: c.enqueueItem,
+	smartCloneController, err := controller.New("smartclone-controller", mgr, controller.Options{
+		Reconciler: reconciler,
 	})
-
-	// Set up an event handler for when PVC resources change
-	// handleObject function ensures we filter PVCs not created by this controller
-	pvcInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
-		AddFunc: c.enqueueItem,
-		UpdateFunc: func(old, new interface{}) {
-			c.enqueueItem(new)
-		},
-		DeleteFunc: c.enqueueItem,
-	})
-
-	return c
+	if err != nil {
+		return nil, err
+	}
+	if err := addSmartCloneControllerWatches(mgr, smartCloneController); err != nil {
+		return nil, err
+	}
+	return smartCloneController, nil
 }
 
-//ProcessNextItem ...
-func (c *SmartCloneController) ProcessNextItem() bool {
-	key, shutdown := c.queue.Get()
-	if shutdown {
-		return false
+func addSmartCloneControllerWatches(mgr manager.Manager, importController controller.Controller) error {
```

Comment:
`importController`

Reply:
Whoops, I thought I had gotten all of those. Fixed.